### PR TITLE
Add: some helpers to compile on ODROID-XU4/XU3

### PR DESCRIPTION
**Pull request by AreaScout**
As the title says, this adds some helpers to successfully compile on ODROID-XU4, the user still has to comment out that ``glCopyImageSubDataOES`` in ``/usr/include/GLES2/gl2ext.h`` but that's about it

I don't know another way around it 👀

### CODE REVIEW

**Comment by hrydgard** [comment]
Not sure I like the approach but also not sure of alternatives...  having to hack gl2ext.h is pretty nasty, there should be better ways to solve that.

File: ext/native/thin3d/GLRenderManager.cpp [hunk 1]
```diff
@@ -806,7 +806,9 @@ void *GLRBuffer::Map(GLBufferStrategy strategy) {
 			if (!hasStorage_) {
 				GLbitfield storageFlags = access & ~(GL_MAP_INVALIDATE_BUFFER_BIT | GL_MAP_FLUSH_EXPLICIT_BIT);
 #ifdef USING_GLES2
+#ifndef ODROID
```
Comment (unknownbrackets) [inline ext/native/thin3d/GLRenderManager.cpp]: Why is this needed?  Don't we get this defined from the header?

-[Unknown]

File: ext/native/gfx_es2/gl3stub.h [hunk 2]
```diff
@@ -504,8 +504,10 @@ extern GL_APICALL void           (* GL_APIENTRY glBindFragDataLocationEXT) (GLui
 extern GL_APICALL GLint          (* GL_APIENTRY glGetProgramResourceLocationIndexEXT) (GLuint program, GLenum programInterface, const GLchar *name);
 extern GL_APICALL GLint          (* GL_APIENTRY glGetFragDataIndexEXT) (GLuint program, const GLchar *name);
 
+#ifndef ODROID
```
Comment (unknownbrackets) [inline ext/native/gfx_es2/gl3stub.h]: Let's change this to `#ifndef GL_EXT_buffer_storage` instead (all 3 times.)

-[Unknown]

File: ext/native/gfx/gl_common.h [hunk 3]
```diff
@@ -4,6 +4,9 @@
 #include <OpenGLES/ES3/gl.h>
 #include <OpenGLES/ES3/glext.h>
 #elif defined(USING_GLES2)
+#ifdef ODROID
```
Comment (unknownbrackets) [inline ext/native/gfx/gl_common.h]: Maybe we can just always define `GL_GLEXT_PROTOTYPES` (ignoring any ODROID define)?  It won't be great for any static PPSSPP linux builds, but should be fine elsewhere I think?

-[Unknown]

File: CMakeLists.txt [hunk 4]
```diff
@@ -86,6 +94,10 @@ else()
   add_definitions(-DNO_VULKAN)
 endif()
 
+if(ODROID)
+  set(VULKAN OFF)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Eventually we may want to remove this define.  If it's a matter of an undefined reference or something, maybe we can just ifdef something more specific.  I know people on Raspberry Pi have had issues too, so I'd rather not laser-focus on ODROID only.

Is it just errors like this?
```
VulkanImage.cpp:89:11: error: cannot convert ‘std::nullptr_t’ to ‘VkImage {aka long long unsigned int}’ in assignment
    image_ = nullptr;
```

If so, these are definitely fixable.  That error is fixed, but in that case the correct code would be `image_ = VK_NULL_HANDLE;`.  It's just a bug in PPSSPP, and it should be fixed.

-[Unknown]

File: CMakeLists.txt [hunk 5]
```diff
@@ -28,6 +28,14 @@ endif()
 if(CMAKE_SYSTEM_PROCESSOR)
 	if(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^arm")
 		set(ARM ON)
+			if(UNIX AND NOT APPLE)
+				execute_process(COMMAND cat /proc/cpuinfo OUTPUT_VARIABLE OUTSTR)
+				string(FIND "${OUTSTR}" "ODROID-XU" pos)
+				if(pos GREATER_EQUAL 0)
+					add_definitions(-mfloat-abi=hard -marm -mtune=cortex-a15.cortex-a7 -mcpu=cortex-a15 -mfpu=neon-vfpv4 -fomit-frame-pointer -ffast-math -DARM_NEON -DODROID)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: A few things here:

1. I don't think `-DARM_NEON` is needed at all.  The compiler defines `__ARM_NEON__` which is what we look for.
2. I'd say we should just add `fast-math` for all gcc/clang, or at least all armv7.  No reason to make it ODROID specific.
3. I know there are a couple armv7 devices lacking NEON, but my preference would be to add `-mfpu=neon-vfpv4` for all armv7 as well.  We could add a cmake option for disabling it if necessary.

Can you run:
```
cmake -E environment
```

Does anything there already contain something that says "hard" or "hardfp"?  It'd be nice to detect hardfp for more than just ODROID.

-[Unknown]

**Review by unknownbrackets — CHANGES_REQUESTED** [review]

**Comment by unknownbrackets** [comment]
Also, for `glCopyImageSubDataOES`, does it work to surround our defs in gl3stub.h with `#ifndef GL_OES_copy_image`?

Not the call - in theory the call should still work, since it's just being defined by the system header instead of us.

-[Unknown]

File: ext/native/thin3d/GLRenderManager.cpp (hunk 1 above)
Comment (AreaScout) [reply ext/native/thin3d/GLRenderManager.cpp]: we get it defined but the Mali Library for T-628 does not define that function, so we would get a linkage error

File: ext/native/gfx_es2/gl3stub.h (hunk 2 above)
Comment (AreaScout) [reply ext/native/gfx_es2/gl3stub.h]: that sound's reasonable, will look into it

File: ext/native/gfx/gl_common.h (hunk 3 above)
Comment (AreaScout) [reply ext/native/gfx/gl_common.h]: I did that, and build failed on Android then, the thing is, each vendor has it's own implementation of this and has it's own headers

File: CMakeLists.txt (hunk 4 above)
Comment (AreaScout) [reply CMakeLists.txt]: I get linkage errors if I let pass it to ``add_definitions(-DNO_VULKAN)``

File: CMakeLists.txt (hunk 5 above)
Comment (AreaScout) [reply CMakeLists.txt]: yes and __ARM_NEON as soon you define ``-mfpu=neon`` that is a left over from the past, I will skip it ;)

**Comment by AreaScout** [comment]
@hrydgard I don't like such hacks too, but as you told already it's hard to get around on that you have your own gl3stub.c which does collide to some header file implementations on SBC's, maybe add your own gl2ext.h or gl3.h ?

**Comment by AreaScout** [comment]
@unknownbrackets 

no surrounding our defs in gl3stub.h for ``glCopyImageSubDataOES`` does not work, it's strange

File: ext/native/thin3d/GLRenderManager.cpp (hunk 1 above)
Comment (unknownbrackets) [reply ext/native/thin3d/GLRenderManager.cpp]: Ah, can we `#define GL_EXT_buffer_storage 1` before including the system GL header then?  That should prevent it from defining its own version, allowing us to do our lookup as normal.

-[Unknown]

File: CMakeLists.txt (hunk 4 above)
Comment (unknownbrackets) [reply CMakeLists.txt]: Which linkage errors?  The only Vulkan code we link with is in this folder:
https://github.com/hrydgard/ppsspp/tree/master/ext/vulkan

So in theory it's all fixable inside PPSSPP's source code.  We're not linking with any system code AFAIK.

-[Unknown]

Comment (AreaScout) [reply CMakeLists.txt]: I will get you postet, rebuilding it atm.

Comment (AreaScout) [reply CMakeLists.txt]: Here it is https://pastebin.com/TQkDwybX

File: ext/native/gfx_es2/gl3stub.h (hunk 2 above)
Comment (AreaScout) [reply ext/native/gfx_es2/gl3stub.h]: changed

File: ext/native/gfx/gl_common.h (hunk 3 above)
Comment (AreaScout) [reply ext/native/gfx/gl_common.h]: changed but I think travis will fail on Android again

File: CMakeLists.txt (hunk 4 above)
Comment (unknownbrackets) [reply CMakeLists.txt]: Hm.  That means that CommonVulkan or at least VulkanMemory.cpp was not linked.  It should still be linked, even though it won't be used.  I don't see any conditionals preventing it from being included in current master:

https://github.com/hrydgard/ppsspp/blob/f87dcefdd89b72953c6b7172ac3e887d9a7f823b/CMakeLists.txt#L400-L410

-[Unknown]

Comment (unknownbrackets) [reply CMakeLists.txt]: Oh, maybe it's some sort of link order problem?  But Common is before native:

https://github.com/hrydgard/ppsspp/blob/f87dcefdd89b72953c6b7172ac3e887d9a7f823b/CMakeLists.txt#L1748

-[Unknown]

File: ext/native/thin3d/GLRenderManager.cpp [hunk 6]
```diff
@@ -806,7 +806,9 @@ void *GLRBuffer::Map(GLBufferStrategy strategy) {
 			if (!hasStorage_) {
 				GLbitfield storageFlags = access & ~(GL_MAP_INVALIDATE_BUFFER_BIT | GL_MAP_FLUSH_EXPLICIT_BIT);
 #ifdef USING_GLES2
+#ifndef GL_EXT_buffer_storage
```
Comment (unknownbrackets) [inline ext/native/thin3d/GLRenderManager.cpp]: This could actually be a bit dangerous - I didn't realize it wasn't defined.

If we define this above `#ifndef GL_GLEXT_PROTOTYPES`, it should prevent gl2ext.h from defining `glBufferStorageEXT` at all, and then we can remove the ifndefs in GLRenderManager and gl3stub.  That should have the least impact on other platforms too.

Sorry, didn't realize before.

-[Unknown]

File: ext/native/gfx/gl_common.h (hunk 3 above)
Comment (unknownbrackets) [reply ext/native/gfx/gl_common.h]: I see.  We could try fighting that battle with `#ifndef GL_NV_framebuffer_blit` in gl3stub, but maybe it is better to use the ODROID define as you suggested.  @hrydgard what do you think?

-[Unknown]

File: CMakeLists.txt (hunk 4 above)
Comment (AreaScout) [reply CMakeLists.txt]: maybe it's an ccache problem, I will rebuild it but not know it has to wait a little bit

**Comment by AreaScout** [comment]
Changed, it's still far from perfect ... you can't believe how often I screwed it up locally, online also 2 times 😁.
I am using the latest headers from khronos now on my ODROID and GL_GLEXT_PROTOTYPES doesn't needs to be defined anymore. (``redeclared as different kind of symbol`` is gone).

``glBufferStorageEXT `` which is not available in libMali.so on ODROIDs is now surrounded by NO_GL_EXT_BUFFER_STORAGE

and lastly I added ARM_NO_VULKAN switch, which can be used for rpi's too 

Let me know what you think ... 

have a good time !

Daniel

**Comment by AreaScout** [comment]
Forgot to say I have also changed this libCommon as added two times now ...

https://github.com/hrydgard/ppsspp/blob/f5c1395df2215f72d238ae01fa000077163601fe/CMakeLists.txt#L1756

**Comment by AreaScout** [comment]
Ok did have some time now, changed ``NO_GL_EXT_BUFFER_STORAGE`` back to ``GL_EXT_buffer_storage`` still all that header file thing confusing me a lot but I have to solve that puzzle. We have to change our gl2ext.h (only for ODROID's) to an vendor specific one now which does not define ``#define GL_EXT_buffer_storage 1`` which was not the case before.

@unknownbrackets can you take a look on it please if this is ok now for you ? Also that ``ARM_NO_VULKAN`` use and the use of this one:

https://github.com/hrydgard/ppsspp/blob/f5c1395df2215f72d238ae01fa000077163601fe/CMakeLists.txt#L1756

File: ext/native/gfx_es2/gl3stub.h [hunk 7]
```diff
@@ -504,8 +504,10 @@ extern GL_APICALL void           (* GL_APIENTRY glBindFragDataLocationEXT) (GLui
 extern GL_APICALL GLint          (* GL_APIENTRY glGetProgramResourceLocationIndexEXT) (GLuint program, GLenum programInterface, const GLchar *name);
 extern GL_APICALL GLint          (* GL_APIENTRY glGetFragDataIndexEXT) (GLuint program, const GLchar *name);
 
+#ifdef GL_EXT_buffer_storage
```
Comment (unknownbrackets) [inline ext/native/gfx_es2/gl3stub.h]: This doesn't really make sense to me though - for other platforms, like probably Android, this will mean that we never use this extension.  Worse, we'll detect it, but then silently not call this function, so buffers will be used but broken.

-[Unknown]

Comment (unknownbrackets) [reply ext/native/gfx_es2/gl3stub.h]: Ah, got it, as you pointed out Android and others do define this.  I guess we end up with the same symbol type.

-[Unknown]

**Comment by AreaScout** [comment]
@hrydgard @unknownbrackets 
do you think we can safely skip using NO_VULKAN definition, it's only used two times (libretro and gpu.cpp) because that seems to be the root cause of any circular dependencies linkage error I have on my ODROID

**Comment by hrydgard** [comment]
I'd love to remove NO_VULKAN.

**Comment by AreaScout** [comment]
Changed, I hope it doesn't screw up Android ... 🤞

**Review by unknownbrackets — APPROVED** [review]
